src/Raven.NewClient/Commands/MultiGetCommand.cs

```diff
@@ -14,66 +15,189 @@ namespace Raven.NewClient.Client.Commands
-                    bool first = true;
+                    var first = true;
                     foreach (var command in commands)
                     {
-                        if (!(first))
+                        if (first == false)
                             writer.WriteComma();
                         first = false;
-                        Context.Write(writer, command);
+                        _context.Write(writer, command);
                     }
                     writer.WriteEndArray();
                 }
             });
 
-            var sb = new StringBuilder($"{node.Url}/databases/{node.Database}/multi_get");
-
-            url = sb.ToString();
+            url = $"{_baseUrl}/multi_get";
 
             return request;
         }
 
-        public override void SetResponse(BlittableJsonReaderObject response)
+        private string GetCacheKey(GetRequest command, out string requestUrl)
+        {
+            requestUrl = $"{_baseUrl}{command.UrlAndQuery}";
+
+            return $"{command.Method}-{requestUrl}";
+        }
+
+        public override void SetResponse(BlittableJsonReaderObject response, bool fromCache)
         {
             if (response == null)
-                throw new InvalidOperationException();
+                ThrowInvalidResponse();
+
+            BlittableJsonReaderArray array;
+            if (response.TryGet("Results", out array) == false || array == null)
+                ThrowInvalidResponse();
+
+            var anyModifications = false;
+            for (int i = 0; i < array.Length; i++)
+            {
+                var result = (BlittableJsonReaderObject)array[i];
+
+                HttpStatusCode statusCode;
+                if (result.TryGet(nameof(Response.StatusCode), out statusCode) == false)
+                    continue;
+
+                if (statusCode != HttpStatusCode.NotModified)
+                    continue;
+
+                var command = _commands[i];
+
+                string requestUrl;
+                var cacheKey = GetCacheKey(command, out requestUrl);
+
+                long cachedEtag;
+                BlittableJsonReaderObject cachedResponse;
+                using (_cache.Get(_context, cacheKey, out cachedEtag, out cachedResponse))
+                {
+                    if (result.Modifications == null)
+                        result.Modifications = new DynamicJsonValue(result);
+
+                    result.Modifications[nameof(Response.Result)] = cachedResponse;
+                    anyModifications = true;
+                }
+            }
+
+            if (anyModifications)
+                response = _context.ReadObject(response, "multi_get/response");
 
             Result = JsonDeserializationClient.BlittableArrayResult(response);
         }
 
+        protected override void CacheResponse(HttpCache cache, RequestExecuterOptions options, string url, HttpResponseMessage response, BlittableJsonReaderObject responseJson)
```

Comment:
You read it all as a single blittable, and that is wrong. What you want to do instead is to break into independent values _as you parse it_, like we do in smuggerl, or @aviviadi work on the BatchHandler.